apps/arweave/src/ar_mining_worker.erl

```diff
@@ -245,362 +284,412 @@ terminate(_Reason, _State) ->
 	Partition2 = ar_node:get_partition_number(RecallRange2Start),
-	Candidate2 = Candidate#mining_candidate{ partition_number2 = Partition2 },
-	Candidate3 = generate_cache_ref(Candidate2),
-	Range1Exists = ar_mining_io:read_recall_range(
-			chunk1, self(), Candidate3, RecallRange1Start),
-	PackingDifficulty = Candidate#mining_candidate.packing_difficulty,
-	State3 = case Range1Exists of
-		true ->
-			Range2Exists = ar_mining_io:read_recall_range(
-					chunk2, self(), Candidate3, RecallRange2Start),
-			case Range2Exists of
-				true -> 
-					State;
+	Candidate2 = generate_cache_ref(Candidate#mining_candidate{ partition_number2 = Partition2 }),
+	%% Check if the recall ranges are readable to avoid reserving cache space for non-existent data.
+	Range1Exists = ar_mining_io:is_recall_range_readable(Candidate2, RecallRange1Start),
+	Range2Exists = ar_mining_io:is_recall_range_readable(Candidate2, RecallRange2Start),
+
+	case {Range1Exists, Range2Exists} of
+		{true, true} ->
+			%% Both recall ranges are readable, so we need to reserve cache space for both.
+			case try_to_reserve_cache_space(2, Candidate2#mining_candidate.session_key, State) of
+				{true, State1} ->
+					%% Read the recall ranges; the result of the read will be reported by the `chunk1` and `chunk2` tasks.
+					ar_mining_io:read_recall_range(chunk1, self(), Candidate2, RecallRange1Start),
+					ar_mining_io:read_recall_range(chunk2, self(), Candidate2, RecallRange2Start),
+					State1;
 				false ->
-					%% Release just the Range2 cache space we reserved with
-					%% try_to_reserve_cache_space/2
-					State2 = update_sub_chunk_cache_size(
-							-ar_block:get_nonces_per_recall_range(PackingDifficulty),
-							SessionKey, State),
-					do_not_cache(Candidate3, State2)
+					%% We don't have enough cache space to read the recall ranges, so we'll try again later.
+					add_delayed_task(self(), computed_h0, Candidate),
+					State
 			end;
-		false ->
-			%% Release the Range1 *and* Range2 cache space we reserved with
-			%% try_to_reserve_cache_space/2
-			update_sub_chunk_cache_size(
-				-(2 * ar_block:get_nonces_per_recall_range(PackingDifficulty)),
-				SessionKey, State)
-	end,
-	{noreply, State3};
+		{true, false} ->
+			%% Only the first recall range is readable, so we need to reserve cache space for it.
+			case try_to_reserve_cache_space(1, Candidate2#mining_candidate.session_key, State) of
+				{true, State1} ->
+					%% Read the recall range; the result of the read will be reported by the `chunk1` task.
+					ar_mining_io:read_recall_range(chunk1, self(), Candidate2, RecallRange1Start),
+					%% Mark chunk2 as missing, not to wait for it to arrive.
+					State2 = mark_chunk2_missing(Candidate2, State1),
+					State2;
+				false ->
+					%% We don't have enough cache space to read the recall range, so we'll try again later.
+					add_delayed_task(self(), computed_h0, Candidate),
```

Comment:
There may be a potential race condition here - but I'm not totally sure it matters. since a miner can sync+pack while mining, it's possible that the first time they get here only recall_range_1 exists, but when the delayed task is processed, both recall_range_1 and recall_range_2 exist. I think this is probably fine? Just calling it out in case.

Maybe be moot if we revert back to delaying the `compute_h0` rather than `computed_h0` task